query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,450 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::btree_map::Entry;
+
+use mentat_core::{
+    Schema,
+};
+
+use mentat_query::{
+    NotJoin,
+    UnifyVars,
+    ContainsVariables,
+};
+
+use clauses::ConjoiningClauses;
+
+use errors::{
+    Result,
+};
+
+use types::{
+    ColumnConstraint,
+    ComputedTable,
+    EmptyBecause,
+};
+
+impl ConjoiningClauses {
+    pub fn apply_not_join(&mut self, schema: &Schema, not_join: NotJoin) -> Result<()> {
+        let projected = match not_join.unify_vars {
+            UnifyVars::Implicit => not_join.collect_mentioned_variables(),
+            UnifyVars::Explicit(vs) => vs.into_iter().collect(),
+        };
+        let mut template = self.use_as_template(&projected);
+        let mut empty_because: Option<EmptyBecause> = None;
+        
+        for clause in not_join.clauses.into_iter() {
+            template.apply_clause(&schema, clause)?;
+            // This is disgusting, but I can't think of anything else right now that works.
+            if template.is_known_empty() {
```

Comment:
In which case, this might be a bug somewhere.

I found that if a clause fills the `empty_because` value when applied to the template, no subsequent clause, even if valid, will populate the `wheres` for the `cc`. Only by removing that `empty_because` value were subsequent valid clauses successfully applied.

I found this when trying to get the `test_only_one_clause_succeeds` test to work.